liblumen_alloc/src/erts/process/priority.rs

```diff
@@ -18,34 +17,58 @@ impl Default for Priority {
 }
 
 impl TryFrom<Atom> for Priority {
-    type Error = Exception;
+    type Error = TryFromAtomError;
 
     fn try_from(atom: Atom) -> Result<Self, Self::Error> {
         match atom.name() {
             "low" => Ok(Priority::Low),
             "normal" => Ok(Priority::Normal),
             "high" => Ok(Priority::High),
             "max" => Ok(Priority::Max),
-            _ => Err(badarg!().into()),
+            name => Err(TryFromAtomError { name }),
         }
     }
 }
 
 impl TryFrom<Term> for Priority {
-    type Error = Exception;
+    type Error = TryFromTermError;
 
     fn try_from(term: Term) -> Result<Self, Self::Error> {
-        term.decode().unwrap().try_into()
+        term.decode()
+            .unwrap()
+            .try_into()
+            .map_err(
+                |try_from_typed_term_error| match try_from_typed_term_error {
+                    TryFromTypedTermError::AtomName(name) => TryFromTermError::AtomName(name),
+                    TryFromTypedTermError::Type => TryFromTermError::Type,
+                },
+            )
     }
 }
 
 impl TryFrom<TypedTerm> for Priority {
-    type Error = Exception;
+    type Error = TryFromTypedTermError;
 
     fn try_from(typed_term: TypedTerm) -> Result<Self, Self::Error> {
         match typed_term {
-            TypedTerm::Atom(atom) => atom.try_into(),
-            _ => Err(badarg!().into()),
+            TypedTerm::Atom(atom) => atom
+                .try_into()
+                .map_err(|TryFromAtomError { name }| TryFromTypedTermError::AtomName(name)),
+            _ => Err(TryFromTypedTermError::Type),
         }
     }
 }
+
+pub struct TryFromAtomError {
+    name: &'static str,
+}
+
+pub enum TryFromTermError {
```

Comment:
We have this type definition in a few places in this PR, let's extract it into the `atom` module and call it `TryAtomFromTermError`. I think we can also get rid of `TryFromTypedTermError` here, as there is no useful difference between the two.